src/main/java/ch/njol/skript/lang/SkriptParser.java

```diff
@@ -263,7 +263,31 @@ private <T extends SyntaxElement> T parse(Iterator<? extends SyntaxElementInfo<?
 								x = x2;
 							}
 							T t = info.c.newInstance();
-							if (t.init(res.exprs, i, getParser().getHasDelayBefore(), res)) {
+							ParserInstance parser = getParser();
+							boolean error = false;
+							List<Class<? extends Structure>> usableStructures = t.getUsableStructures();
+							if (!usableStructures.isEmpty()) {
+								if (parser.getCurrentStructure() == null) {
+									error = true;
+									Skript.error("'" + res.expr + "' cannot be used outside of the allowed structures.");
+								} else if (usableStructures.stream().noneMatch(parser::isCurrentStructure)) {
+									error = true;
+									Skript.error("'" + res.expr + "' cannot be used in " + Utils.a(parser.getCurrentStructure().toString(null, false)) + " structure.");
+								}
+							}
+							List<Class<? extends Section>> usableSections = t.getUsableSections();
+							List<TriggerSection> sections = parser.getCurrentSections();
```

Comment:
This and current could probably be declared after the initial check?

Reply:
been changed c: